include/sparrow/layout/array_registry.hpp

```diff
@@ -0,0 +1,442 @@
+     * auto result = registry.dispatch([](auto&& array) {
+     *     return array.size();
+     * }, *arr_wrapper);
+     */
+    class array_registry
+    {
+    public:
+
+        /// Factory function type that creates an array_wrapper from an arrow_proxy
+        using factory_func = std::function<cloning_ptr<array_wrapper>(arrow_proxy)>;
+
+        /// Extension predicate that checks if a proxy matches an extension type
+        using extension_predicate = std::function<bool(const arrow_proxy&)>;
+
+        /// Visitor result type alias
+        template <class F>
+        using visit_result_t = std::invoke_result_t<F, null_array>;
+
+        /**
+         * @brief Get the singleton registry instance.
+         */
+        [[nodiscard]] SPARROW_API static array_registry& instance();
+
+        /**
+         * @brief Register a base type factory.
+         *
+         * Base type factories are used when no extension matches. They handle
+         * the standard Arrow data types.
+         *
+         * NOTE: This registers the factory for array creation. The dispatch
+         * mechanism (visit/dispatch_base_type) uses a compile-time switch statement
+         * because C++ requires knowing the concrete type at compile time to unwrap
+         * the array_wrapper. While we could theoretically store type-erased dispatch
+         * functions, this would add significant runtime overhead and complexity for
+         * no practical benefit, since the set of base types is fixed and known at
+         * compile time.
+         *
+         * @param dt The data_type enum value
+         * @param factory Factory function to create the array
+         */
+        SPARROW_API void register_base_type(data_type dt, factory_func factory);
+
+        /**
+         * @brief Register an extension type factory.
+         *
+         * Extension types are checked before base types. An extension is selected
+         * when its base_type matches and its predicate returns true for the proxy.
+         *
+         * @param base_type The underlying base data_type
+         * @param extension_name The value of "ARROW:extension:name" metadata
+         * @param factory Factory function to create the array
+         */
+        SPARROW_API void
+        register_extension(data_type base_type, std::string_view extension_name, factory_func factory);
+
+        /**
+         * @brief Register an extension type with custom predicate.
+         *
+         * This overload allows for more complex extension detection logic beyond
+         * simple metadata name matching.
+         *
+         * @param base_type The underlying base data_type
+         * @param predicate Custom function to check if proxy is this extension
+         * @param factory Factory function to create the array
+         */
+        SPARROW_API void
+        register_extension_with_predicate(data_type base_type, extension_predicate predicate, factory_func factory);
```

Comment:
This should be named `register_extension`  too, the compiler can figure out the right overload at the call site.